airflow/models.py

```diff
@@ -119,12 +119,17 @@ def clear_task_instances(tis, session, activate_dag_runs=True):
             if ti.job_id:
                 ti.state = State.SHUTDOWN
                 job_ids.append(ti.job_id)
-        # todo: this creates an issue with the webui tests
-        # elif ti.state != State.REMOVED:
-        #     ti.state = State.NONE
-        #     session.merge(ti)
         else:
-            session.delete(ti)
+            try:
+                ti_retries = dag.get_task(ti.task_id).retries
+                ti.max_tries = ti.try_number + ti_retries
+            except (AttributeError, AirflowException):
```

Comment:
You can use the methods that get_task uses internally instead to avoid having to deal with exceptions:
    if dag_id not in dagbag.dags:
    if not dag.has_task(task_id):

Might be good to factor these into a new method task.exists if you think it makes sense.